### DISCUSSION TITLE
Virtual character lifts off the ground on vertical moving platform

### DISCUSSION THREAD

**Discussion by fredUmlaut** [jrouwe/JoltPhysics#537] [Q&A] [answered]
Reproduction:
1. Change the vertical movement of "mVerticallyMovingBody" in CharacterVirtualTest.cpp from a sinusoidal to a linear up/down movement
2. Place a virtual character controller on top
3. When "mVerticallyMovingBody" changes from upward to downward motion the character lifts off the ground

The virtual character should not get an upward impulse when the platform changes vertical direction, but should stay on the platform.

**Comment by jrouwe** [comment]
Can you try #532?

**Comment by fredUmlaut** [comment]
Thanks for the update. 
I tried with #532 but the issue is not fixed.
I move the platform with a constant offset of 0.04 each frame, and the character keeps lifting
off when the platform direction is changed from upwards to downwards.
![jolt](https://github.com/jrouwe/JoltPhysics/assets/14937735/0bd88b09-93b9-4a07-99a6-c869cd15c72e)

**Comment by jrouwe** [comment]
Ah, now I see what you mean (I misread your description and saw a different issue). 

This is actually what would happen in real life if you were able to instantly switch a lift from going up at 2.4m/s (= 0.04 * 60) to going down at 2.4m/s. The character has inertia and since it is not attached to the platform it will continue going up at 2.4 - 0.5 * 9.81 * t m/s. So only after about 1 second it has enough down velocity to reach the 2.4 m/s lift velocity and it starts catching up again. 

If you don't want this you can implement your own rules for the character and update its velocity instantly. In the sample, you can do so by changing [this](https://github.com/jrouwe/JoltPhysics/blob/53122918cef2e13091ded75eb1055ff14d8b4639/Samples/Tests/Character/CharacterVirtualTest.cpp#L122-L133) code to:

```
	if (mCharacter->GetGroundState() == CharacterVirtual::EGroundState::OnGround) // If on ground
	{
		// Assume velocity of ground when on ground
		new_velocity = ground_velocity;

		// Jump
		if (inJump && (current_vertical_velocity.GetY() - ground_velocity.GetY()) < 0.1f) // And not moving away from ground
			new_velocity += sJumpSpeed * mCharacter->GetUp();
	}
	else
		new_velocity = current_vertical_velocity;
```

**Comment by fredUmlaut** [comment]
Thank for the quick reply and sorry for being unclear on the issue description.

Thanks for the code fragment! I think this is what most users of a virtual character controller are looking for, otherwise 
they would use a normal physics body.

As in this sample the vertical displacement on moving platforms isn't bound to inertia I think it would be good to make the horizontal movement also disregard inertia by default. A define could be added to use the current behavior.
movement.

Thanks again for all your work on this!

**Comment by jrouwe** [comment, marked as answer]
Getting rid of horizontal inertia is also easy, just change [this](https://github.com/jrouwe/JoltPhysics/blob/53122918cef2e13091ded75eb1055ff14d8b4639/Samples/Tests/Character/CharacterVirtualTest.cpp#L98) line to:

```mDesiredVelocity = sCharacterSpeed * inMovementDirection;```

Both changes are in what you would implement in your own game code (they're both in the demo application), so I don't really see the need for defines. The demo is just an example of how you could use it, not how you have to use it.

**Comment by fredUmlaut** [comment]
Thanks for the details!
